Complete Typescript support

reflog opened the pull request:
I'd like to create a complete typescript support for Cerebral.
I'll go through all the packages, in each create typings folder and reference it from package.json and will add a typings test.
Will close #796 and hopefully supersede https://github.com/cerebral/cerebral/pull/808

Guria commented on .gitignore:
```diff
@@ -12,3 +12,4 @@ website/builder/docs
 **/yarn.lock
 **/yarn-error.log
 !/yarn.lock
+.vscode/launch.json
```
it's better add environment specific ignore into your local gitignore rather than in projects repos.

Guria commented on packages/cerebral/typings/__test__/examples/code1.ts:
```diff
@@ -0,0 +1,13 @@
+/// <reference path="../../index.d.ts" />
```
do we really need referencing using this syntax?  It looks very odd to me

Guria commented on packages/cerebral/typings/index.d.ts:
```diff
@@ -0,0 +1,109 @@
+// <reference types="function-tree" />
+
+
+declare module 'cerebral/devtools' {
```
I don't think we should use ambient module definitions.

Guria replied:
see eg redux definitions. it doesn't use declare.

reflog replied:
Are you sure about this? Redux is an easier case - they just have function and no sub-modules... Maybe I'm wrong, but I can't find an example project that has a structure more similar to Cerebral's

reflog replied:
Nope. Can't find a way to handle `'cerebral/devtools'` properly. Suggestions welcome.

Guria replied:
you just place an d.ts file for each entry point with the same name. typescript compiler itself will resolve it

reflog replied:
That's not that simple. Unless I put index.d.ts in our root folder (/packages/cerebral/), TS compiler doesn't recognize it as 'cerebral' module and doesn't allow `import { Controller } from "cerebral";`. 

So if I move index.d.ts to src folder (in order to work with /devtools/ subfolder) the compilation breaks.

Also, another problem will come with inferno and react 'submodules' inside viewFactories folder, since they are not folders, so I'm not sure how to place d.ts files there.

Help! :)

reflog commented:
All great points, will take care of it

On Tue, May 9, 2017, 20:28 Aleksey Guryanov <notifications@github.com>
wrote:

> *@Guria* commented on this pull request.
> ------------------------------
>
> In packages/cerebral/typings/index.d.ts
> <https://github.com/cerebral/cerebral/pull/892#discussion_r115550451>:
>
> > @@ -0,0 +1,109 @@
> +// <reference types="node" />
> +// <reference types="function-tree" />
> +
> +
> +declare module 'cerebral/devtools' {
>
> see eg redux definitions. it doesn't use declare.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cerebral/cerebral/pull/892#discussion_r115550451>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGtNORw0eqdF3aWlTP8YW61jf26s5dGks5r4KI5gaJpZM4NVPC0>
> .
>

Guria commented on .cz-config.js:
```diff
@@ -12,6 +12,7 @@ module.exports = {
     {value: 'test', name: 'test:     Adding missing tests'},
     {value: 'chore', name: 'chore:    Changes to the build process or auxiliary tools\n            and libraries such as documentation generation'},
     {value: 'revert', name: 'revert:   Revert to a commit'},
+    {value: 'ts', name: 'ts:   Typescript/Typings retlated changes'},
```
retlated -> related

Guria commented on .gitignore:
```diff
@@ -11,4 +11,4 @@ website/builder/docs
 **/yarn.lock
 **/yarn-error.log
-!/yarn.lock
+!/yarn.lock
```
please keep newlines at the end of all files

Guria commented on packages/cerebral/devtools.d.ts:
```diff
@@ -0,0 +1,13 @@
+export interface DevOptions {
```
Do we need to export this interface? Take a note that all exported interfaces becomes part of public API and it will be covered by semver. Eg,, renaming exported interface or any other breaking chage in it would require major bump.

Guria commented on packages/cerebral/index.d.ts:
```diff
@@ -0,0 +1,35 @@
+// <reference types="node" />
```
what this stands for?

Guria commented on packages/cerebral/react.d.ts:
```diff
@@ -0,0 +1,9 @@
+
+}
+
+export class Container {
```
shouldn't it be extended from React.Component?

Guria commented on packages/cerebral/tags.d.ts:
```diff
@@ -0,0 +1,6 @@
+import { IPath } from "../function-tree/index";
```
from 'function-tree'

Guria replied:
why this left?

Guria replied:
Ping

Guria commented on packages/cerebral/typings/__test__/examples/code1.ts:
```diff
@@ -0,0 +1,12 @@
+// if i try importing from 'cerebral' and 'cerebral/devtools' , it fails to compile
```
let's remove this comment.
it's expected behavior. js tests also requires using relative pathes.
btw, why this file burried so deep? can we collocate tests near d.ts files?

Guria commented on packages/cerebral/typings/__test__/examples/code1.ts:
```diff
@@ -0,0 +1,12 @@
+// if i try importing from 'cerebral' and 'cerebral/devtools' , it fails to compile
+import { Controller } from "../../../index";
```
index could be ommited

Guria commented on packages/cerebral/typings/__test__/examples/code1.ts:
```diff
@@ -0,0 +1,12 @@
+});
+
+
+controller.getSignal("ASD");
```
should we have a wider set of API used?

Guria commented on packages/forms/index.d.ts:
```diff
@@ -0,0 +1,6 @@
+import { IPath } from "../function-tree/index";
```
from 'function-tree'

Guria commented on packages/forms/operators.d.ts:
```diff
@@ -0,0 +1,5 @@
+import { IPath } from "../function-tree/index";
```
from 'function-tree'

Guria commented on packages/function-tree/index.d.ts:
```diff
@@ -0,0 +1,141 @@
+/// <reference types="node" />
+
+export interface IPayload {
```
Lets make sure we are exporting here only needed interfaces.

Guria replied:
Let also avoid using I prefix for interface names as well.

Guria commented on packages/function-tree/index.d.ts:
```diff
@@ -0,0 +1,141 @@
+export function ExecutionProvider(execution: any, abort: Abort): Provider
+export function ContextProvider(extendedContext: any): Provider
+
+export class FunctionTree extends NodeJS.EventEmitter {
```
we extend eventemitter3. which seems to [have own typings](https://github.com/primus/eventemitter3/blob/master/index.d.ts) why we declare here NodeJS.EventEmitter?

Guria commented on packages/http/index.d.ts:
```diff
@@ -0,0 +1,7 @@
+import { Provider } from "../function-tree/index"
```
from 'function-tree'

reflog commented:
ok, addressed all comments except the location of the typescript tests. please propose a structure, because I need both the test runner and a folder with .ts files that will execute the tests and I don't want to litter the codebase too much..

Guria commented:
sorry. just found pending review comments

Guria commented on packages/cerebral/package.json:
```diff
@@ -35,5 +38,10 @@
+  },
+  "devDependencies": {
+    "@types/node": "^7.0.18",
+    "typescript": "^2.3.2",
```
all dev dependencies should be defined only in monorepo package.json

Guria commented:
about tests location. let put them in one folder deep, eg, `ts`:
```
ts/
  typescript.spec.js
  some-file-to-compile.ts
  another-file-to-compile.ts
lib/
src/
index.js
index.d.ts
operators.js
operators.d.ts
```

henri-hulski commented:
@reflog you should not merge master but rebase on master. 
Like this this the PR cannot be merged back. 
If you need help I can give you detailed instructions.

reflog commented:
crap! i thought that's what I did! I ran `git rebase origin/master` in my branch,...

henri-hulski commented:
Strange. That sounds correct. I'm using `git rebase master` on the branch after pulling to master, but this looks even faster. 
But here definitely happens a merge.

henri-hulski commented:
After you need to do `git push --force`.

Guria commented:
@reflog will you resolve rest of comments?

reflog commented:
@Guria - I've the commit up.

gaspard commented on packages/function-tree/package.json:
```diff
@@ -46,8 +46,5 @@
       "**/*.test.js",
       "**/testHelper.js"
     ]
-  },
```
The package `mock-socket` is required to run tests in function-tree. Please leave the devDependencies where they are needed. Most of them are missing in other packages and this is wrong and we will need to fix it. I would like to setup a rule where dependencies not in any package's `devDependencies` or `dependencies` will  get removed from main package.json at some point.

reflog replied:
done

gaspard commented on packages/cerebral/package.json:
```diff
@@ -35,8 +35,5 @@
     ]
-  },
-  "devDependencies": {
-    "mock-socket": "^6.0.4"
```
`mock-socket` does not seem to be used in cerebral testing so this is good cleanup.

gaspard commented:
What is blocking this PR ? Can we move forward ? It would really be good to clear pull requests so that we can finish transition to "no lerna" thing and work on release and build issues.
